examples/graph/gat_node_classification.py

```diff
@@ -0,0 +1,398 @@
+
+            loss = self.compiled_loss(labels, tf.gather(outputs, indices))
+
+        grads = tape.gradient(loss, self.trainable_weights)
+
+        optimizer.apply_gradients(zip(grads, self.trainable_weights))
+
+        self.compiled_metrics.update_state(labels, tf.gather(outputs, indices))
+
+        return {m.name: m.result() for m in self.metrics}
+
+    def predict_step(self, data):
+
+        indices = data
+
+        outputs = self([self.node_features, self.adjacency])
+
+        return tf.nn.softmax(tf.gather(outputs, indices))
+
+    def test_step(self, data):
+
+        indices, labels = data
+
+        outputs = self([self.node_features, self.adjacency])
+
+        loss = self.compiled_loss(labels, tf.gather(outputs, indices))
+
+        self.compiled_metrics.update_state(labels, tf.gather(outputs, indices))
+
+        return {m.name: m.result() for m in self.metrics}
+
+
+"""
+### Train and predict
+"""
+
+# Define hyper-parameters
+HIDDEN_UNITS = 100
+NUM_HEADS = 8
+NUM_LAYERS = 3
+OUTPUT_DIM = len(class_values)
+
+NUM_EPOCHS = 100
+BATCH_SIZE = 256
+VALIDATION_SPLIT = 0.1
+LEARNING_RATE = 3e-1
+MOMENTUM = 0.9
+
+loss_fn = keras.losses.SparseCategoricalCrossentropy(from_logits=True)
+optimizer = keras.optimizers.SGD(LEARNING_RATE, momentum=MOMENTUM)
+accuracy_fn = keras.metrics.SparseCategoricalAccuracy(name="acc")
+early_stopping = keras.callbacks.EarlyStopping(
+    monitor="val_acc", min_delta=0, patience=20, restore_best_weights=True
+)
+
+# Build model
+gat_model = GraphAttentionNetwork(
+    node_features, adjacency, HIDDEN_UNITS, NUM_HEADS, NUM_LAYERS, OUTPUT_DIM
+)
+
+# Compile model
+gat_model.compile(loss=loss_fn, optimizer=optimizer, metrics=[accuracy_fn])
+
+
+gat_model.fit(
+    x=train_indices,
+    y=train_labels,
+    validation_split=VALIDATION_SPLIT,
+    batch_size=BATCH_SIZE,
+    epochs=NUM_EPOCHS,
+    callbacks=[early_stopping],
+    verbose=2,
+)
+
+_, test_accuracy = gat_model.evaluate(x=test_indices, y=test_labels, verbose=0)
+
+print("--" * 38 + f"\nTest Accuracy {test_accuracy*100:.1f}%")
+
+"""
+### Predict (probabilities)
+"""
+test_probs = gat_model.predict(x=test_indices)
+
+mapping = {v: k for (k, v) in class_idx.items()}
+
+for i, (probs, label) in enumerate(zip(test_probs[:10], test_labels[:10])):
+    print(f"Example {i+1}: {mapping[label]}")
+    for j, c in zip(probs, class_idx.keys()):
+        print(f"\tProbability of {c: <24} = {j*100:7.3f}%")
+    print("---" * 20)
+
+"""
+## Conclusions
+
+Results look OK! The GAT model seems to correctly predict the subjects of the papers,
+based on what they cite, about 80-85% of the time. Further improvements could possible be
+made in finetuning the hyperparameters of the GAT. For instance, the number of layers
+and hidden units, optimizer/learning rate, dropout, or to modify the preprocessing step.
+We could also try to implement *self-loops*
+(i.e., paper X cites paper X) and/or make the graph *undirected*.
```

Comment:
What's the best accuracy you achieve when not restricted by the limitation of keeping the runtime short?

Reply:
I haven't yet investigated this enough to have a good answer (also notice, I haven't set a random seed for the train/test split, so the results have varied based on that). This is actually the first time I implement a GAT for node classification, before I only used it for graph classification (e.g., predicting molecular properties). 

To the best of my knowledge (based on my experiments offline), I would say that you could likely squeeze out a 3-5% increase in accuracy with relatively straight forward modifications (adding layers/units, regularization, or adding a learning rate schedule, etc.)

About limitations on runtime: I think the data set is small enough, and the model efficient enough, to not be that limited by #layers/units etc.